### PR TITLE
Packaging bifold as a library

### PR DESCRIPTION
**Pull request by cvarjao**
# Summary of Changes
- isolating mobile app boilerplate in its own folder: This is effective a minimal template of a what a bifold-based app would need to have
   - move files from `android` to `bifold-app/android`
   - move files from `ios` to `bifold-app/ios`
   - move `index.js` to `bifold-app/index.js`
   - move `App.tsx` to `bifold-app/App.tsx`
- introduce configuration Context
-  move all `dependencies` to `peerDepedencies`: This might need to be revisited in the future. Most of react native dependencies only works if it is a direct dependency of the project.
- add react-native-builder-bob for packaging as a library
- Change to Github action to reflect the folder changes

# TODO

- [x] Remove `yarn.lock` file
- [x] Fix native build action in the next PR

# Related Issues

Please reference here any issue #'s that are relevant to this PR, or simply enter "N/A" if this PR does not relate to any existing issues.

# Pull Request Checklist

Tick all boxes below to demonstrate that you have completed the respective task. If the item does not apply to your this PR **check it anyway** to make it apparent that there's nothing to do.

- [x] All commits contain a DCO `Signed-off-by` line (we use the [DCO GitHub app](https://github.com/apps/dco) to enforce this);
- [x] Updated LICENSE-3RD-PARTY.md for any added dependencies or vendored components;
- [x] Updated documentation as needed for changed code and new or modified features;
- [x] Added sufficient [tests](../__tests__/) so that overall code coverage is not reduced.

If you have _any_ questions to _any_ of the points above, just **submit and ask**! This checklist is here to _help_ you, not to deter you from contributing!

Pro Tip 🤓

- Read our [contribution guide](../CONTRIBUTING.md) at least once; it will save you a few review cycles!
- Your PR will likely not be reviewed until all the above boxes are checked and all automated tests have passed.

_PR template adapted from the Python attrs project._

### CODE REVIEW

File: App/index.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+/* eslint-disable import/no-cycle */
+export { DispatchAction } from './store/reducer'
+
+export { InfoTextBox, CheckBoxRow } from './components'
+
+export { default as Button, ButtonType } from './components/buttons/Button'
+
+export { testIdWithKey } from './utils/testable'
+
+export { default as ConnectionModal } from './components/modals/ConnectionModal'
```
Comment (amanji) [inline App/index.ts]: In line with previous comments we can simply import default modules using:
`import <module> from <path>`

File: App/utils/configurationContext.tsx [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+const ConfigurationContext = createContext<ConfigurationContext>(null as unknown as ConfigurationContext)
+export const ConfigurationProvider = ConfigurationContext.Provider
+
+export const useConfigurationContext = () => useContext(ConfigurationContext)
+
+interface ConfigurationProviderProps {
+  value: ConfigurationContext
+}
+/*
```
Comment (amanji) [inline App/utils/configurationContext.tsx]: Is this meant to be commented out for testing?

File: bifold-app/App.tsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  defaultTheme as theme,
+  ConfigurationContext,
+  ConfigurationProvider,
+  initLanguages,
+  defaultTranslationResources,
+  defaultOnboardingPages as onboardingPages,
+  defaultSplashScreen as splashSreen,
+  defaultTerms as termsSreen,
+} from 'aries-bifold'
+import { default as React, useEffect, useState } from 'react'
```
Comment (amanji) [inline bifold-app/App.tsx]: `import React, { useEffect, useState } from 'react'`

File: bifold-app/App.tsx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  AutoAcceptCredential,
+  ConsoleLogger,
+  HttpOutboundTransport,
+  LogLevel,
+  MediatorPickupStrategy,
+  WsOutboundTransport,
+} from '@aries-framework/core'
+import AgentProvider from '@aries-framework/react-hooks'
+import { agentDependencies } from '@aries-framework/react-native'
+import {
```
Comment (amanji) [inline bifold-app/App.tsx]: I'm not sure if I agree with mixing component imports with configurations from the same path. Can we split these up or package them better for re-use as a lib?

**Comment by amanji** [comment]
@cvarjao there's a yarn lockfile. Should that be removed?

File: bifold-app/App.tsx (hunk 3 above)
Comment (cvarjao) [reply bifold-app/App.tsx]: Fixed

File: bifold-app/App.tsx (hunk 4 above)
Comment (cvarjao) [reply bifold-app/App.tsx]: I am not sure what this means. But I agree that we can refactor the packaging better after. I am thinking we can introduce namespaces and export them to keep a good separation between interfaces and classes.

File: App/utils/configurationContext.tsx (hunk 2 above)
Comment (cvarjao) [reply App/utils/configurationContext.tsx]: removed.

**Comment by jleach** [comment]
@cvarjao Able to get the automated tests working? If you check the Action's logs they'll tell you what needs to be done.

**Review by JamesKEbert — DISMISSED** [review]
This looks good! I think I'd be interested in seeing if we could adjust the dependency issue as you mentioned, as it'd definitely make using it simpler/cleaner.

**Comment by cvarjao** [comment]
@JamesKEbert , I including a link about "linking of dependencies" for future reference:
https://github.com/react-native-community/cli/issues/1347
There is an issue of ending up with different version of the dependency

Someone suggested a workaround/patch for this issue:
https://stackoverflow.com/a/64097197/3377438

File: .github/workflows/main.yaml [hunk 5]
```diff
@@ -55,25 +55,26 @@ jobs:
       # git for cache keys.
       - name: Generate cache key
         run: |
-          echo $(git rev-parse HEAD:App) > ./dd-cache-key.txt
-          echo $(git rev-parse HEAD:ios) >> ./dd-cache-key.txt
+          echo $(git rev-parse HEAD:core/App) > ./dd-cache-key.txt
```
Comment (jleach) [inline .github/workflows/main.yaml:58]: There isn't anything useful in core/ to cache here. It's all in `ios`.

**Review by jleach — DISMISSED** [review]
Checked out, ran locally, works as expected.

File: .github/workflows/main.yaml (hunk 5 above)
Comment (cvarjao) [reply .github/workflows/main.yaml:58]: I was just maintaining compatibility with what already was there. Can this be addressed later?

Comment (jleach) [reply .github/workflows/main.yaml:58]: Yup

**Review by jleach — APPROVED** [review]
Tested as a group. Able to accept an connection, offer, and proof. LGTM.

**Pull request merged**